tests/src/whisk/core/database/test/MultipleReadersSingleWriterCacheTests.scala

```diff
@@ -0,0 +1,215 @@
+        for (i <- 1 to nIters) {
+            doCRUD("CONCURRENT CRUD to shared keys, with update first", sharedKeys, 1000, false)
+                .nWriteInhibits.get should be(0)
+        }
+    }
+
+    def sharedKeys = { i: Int => "foop_" + (i % 2) }
+
+    def doCRUD(
+        testName: String,
+        key: Int => String,
+        delay: Int = 1000,
+        readsFirst: Boolean = true,
+        nThreads: Int = 10): Inhibits = {
+
+        System.out.println();
+        System.out.println(testName);
+
+        val exec = Executors.newFixedThreadPool(nThreads)
+        val inhibits = Inhibits()
+
+        for (i <- 1 to nThreads) {
+            exec.submit(new Runnable { def run() = { doReadWriteRead(key(i), inhibits, readsFirst).go(delay) } })
+        }
+
+        exec.shutdown
+        exec.awaitTermination(2, TimeUnit.MINUTES)
+
+        inhibits.debug(this)
+        inhibits
+    }
+
+    case class Inhibits(
+        nReadInhibits: AtomicInteger = new AtomicInteger(0),
+        nWriteInhibits: AtomicInteger = new AtomicInteger(0)) {
+
+        def debug(from: AnyRef) {
+            logger.debug(from, "InhibitedReads: " + nReadInhibits);
+            logger.debug(from, "InhibitedWrites: " + nWriteInhibits);
+        }
+
+        def someInhibits: Boolean = { nReadInhibits.get > 0 || nWriteInhibits.get > 0 }
+    }
+
+    private case class doReadWriteRead(key: String, inhibits: Inhibits = Inhibits(), readFirst: Boolean = true) {
+        def go(implicit delay: Int): Inhibits = {
+            val latch = new CountDownLatch(2)
+
+            implicit val transId = TransactionId.testing
+
+            if (!readFirst) {
+                // we want to do the update before the first read
+                cacheUpdate(key, key, delayed("bar_b")) onFailure {
+                    case t =>
+                        inhibits.nWriteInhibits.incrementAndGet();
+                }
+            }
+
+            cacheLookup(key, delayed("bar"), true) onComplete {
+                case Success(s) => {
+                    latch.countDown()
+                }
+                case Failure(t) => {
+                    latch.countDown()
+                    inhibits.nReadInhibits.incrementAndGet();
+                }
+            }
+
+            if (readFirst) {
+                // we did the read before the update, so do the write next
+                cacheUpdate(key, key, delayed("bar_b")) onFailure {
+                    case t =>
+                        inhibits.nWriteInhibits.incrementAndGet();
+                }
+            }
+
+            cacheLookup(key, delayed("bar_c"), true) onComplete {
+                case Success(s) => {
+                    latch.countDown();
+                }
+                case Failure(t) => {
+                    inhibits.nReadInhibits.incrementAndGet();
+                    latch.countDown();
+                }
+            }
+
+            latch.await(2, TimeUnit.MINUTES)
+
+            inhibits
+        }
```

Comment:
Does this testcase assume, that the calls are running in the order they are put? It reads a bit like that.